src/wasm/wasm-type.cpp

```diff
@@ -3110,23 +3106,8 @@ buildNominal(std::vector<std::unique_ptr<HeapTypeInfo>> infos) {
   auto start = std::chrono::steady_clock::now();
 #endif
 
-  // Move the HeapTypes and the Types they reach to the global stores. First
-  // copy reachable temporary types into the global type store and replace their
-  // uses in the HeapTypeInfos. Then move all the HeapTypeInfos into the global
-  // store. We have to copy types first because correctly hashing the
-  // HeapTypeInfos depends on them reaching only canonical types. That also
-  // means we can't reuse `globallyCanonicalize` here.
-  Locations locations;
-  for (auto& info : infos) {
-    HeapType root = asHeapType(info);
-    locations.walkRoot(&root);
-  }
-  globallyCanonicalizeTypes(locations);
-
-  std::vector<HeapType> heapTypes;
-  for (auto& info : infos) {
-    heapTypes.push_back(globalHeapTypeStore.insert(std::move(info)));
-  }
+  // Move the HeapTypes and the Types they reach to the global stores.
+  std::vector<HeapType> heapTypes = globallyCanonicalize(infos);
```

Comment:
I don't have any performance concerns. `globallyCanonicalize` calculates type locations, acquires the lock on the `globalHeapTypeStore`, moves all the HeapTypes into the store, then moves all the reachable types into their store. The removed code below did all of that as well except for take the lock. In practice the lock doesn't matter because module parsing is single-threaded.

Reply:
Makes sense. Maybe not important to benchmark here then.